### ISSUE TITLE
[Feature Request] HParams Class of TensorFlow

### ISSUE THREAD

**Issue by seungwonpark** [pytorch/pytorch#16838]
## 🚀 Feature
>Class to hold a set of hyperparameters as name-value pairs.

cited from tensorflow doc.

## Motivation
[`HParams`](https://www.tensorflow.org/api_docs/python/tf/contrib/training/HParams) class in Tensorflow is useful for hyperparameter management, but I couldn't find similar one in PyTorch. I will be grateful if someone points out already-existing feature.

## Pitch
How about adopting [TF dependecy-removed version of `HParams`](https://github.com/r9y9/deepvoice3_pytorch/blob/master/hparam_tf/hparam.py)? 
I'm not sure whether TF's license and PyTorch's license will conflict or not.

## Alternatives

We may use json file for hparam management like [this](https://github.com/victoresque/pytorch-template), but then we can't expect some rich features of `HParams` class in TF.

What will be the best practice for managing hparams in PyTorch without `HParams`?

**Comment by Kaixhin** [comment]
What are the important features from `HParams` that you see missing from the native [`argparse`](https://docs.python.org/3/library/argparse.html) module?

**Comment by seungwonpark** [comment]
(I'm sorry for being late.)
1. There is no way to directly merge `dict` or `json` to `argparse` parser.
Hence, if we want to use both `argparse` and json file, we can't use single object.
`HParams` have a method for exporting/importing json or protobuf.
2. (This can be controversial) In my opinion, using only `argparse` takes long time to code, compared to using `HParams` or json.

## Using only `argparse`
```python
parser.add_argument('--filters', type=int, default=256)
parser.add_argument('--blocks', type=int, default=512)
parser.add_argument('--nonlinear', type=str, default='relu')
parser.add_argument('--sample_rate', type=int, default=16000)
parser.add_argument('--hop_length', type=int, default=200)
```

## Using `HParams`
```python
hp = tf.contrib.training.HParams(
    filters=256,
    blocks=512,
    nonlinear='relu',
    sample_rate=16000,
    hop_length=200,
)
```

**Comment by fmassa** [comment]
In the first case, one option is to use [ConfigParser](https://docs.python.org/3/library/configparser.html) in combination with `argparse`, [here is one example](http://blog.vwelch.com/2011/04/combining-configparser-and-argparse.html)

**Comment by Kaixhin** [comment]
I think this level of functionality is probably a better fit for a higher level wrapper for PyTorch, like [Ignite](https://pytorch.org/ignite/). What do you think @alykhantejani?

**Comment by Neabfi** [comment]
That would be a first step to support hparams in tensorboard. This is a really important missing feature.


<img width="1225" alt="Screen Shot 2019-05-20 at 16 59 07" src="https://user-images.githubusercontent.com/8831940/58031415-a9d52980-7b20-11e9-9fa2-7c7f2d1a15cc.png">

**Comment by ezyang** [comment]
cc @lanpa since apparently this is TensorBoard related

**Comment by ezyang** [comment]
@gchanan: It's really not clear if this is appropriate to put in PyTorch, or just some higher level wrapper. There doesn't seem to be any real connection to the rest of PyTorch. Is there a parameter sweep? Can I pass it to an optimizer?

**Comment by mbforbes** [comment]
+1 including the tensorboard hparams wrapper in the current tensorboard support would be really helpful! Right now it seems the best support for viewing hparams search in pytorch is checking out a custom fork of tensobaordX which adds support (https://github.com/lanpa/tensorboardX/issues/381). At least for me, supporting tf's view of hparam specs and merging configs isn't as important—I agree it could be done in a higher level wrapper—as adding hparams export support to a visualizer.

**Comment by lanpa** [comment]
https://github.com/lanpa/tensorboardX/pull/470

@ezyang The plugin is definitely not pytorch specific. Indeed, the only thing needed by the plugin are two dictionaries describing the hyper parameters and the result of that set of parameter.
`writer.add_hparam(hparam_dict= {'lr': 0.1, 'bsize': 12},
          metrics= {'accuracy': 0.987, 'loss': 10})`

also cc @orionr for further concerns about whether any pytorch specific feature can be added before putting into torch.tensorboard.

**Comment by orionr** [comment]
I think there's opportunities to connect this with parameter sweeping in the core library, but also just having better parity with TensorBoard support in TensorFlow is a good thing, so I'd support this. It should be a simple `add_hparam` or equivalent call like @lanpa noted.

cc @natalialunova @sanekmelnikov

**Comment by soumith** [comment]
from what I see, we really dont need hparams just to suppor the TensorBoard integration.
We need the tensorboard wrapper to have a caster from a dict to a HParam (if TB requires a HParam object specifically)

**Comment by orionr** [comment]
For minimum support, we really want to just have a new `add_hparam` method. This is matching https://github.com/lanpa/tensorboardX/pull/470/files#diff-0900f72201599ce7d24575f21ca88a85R325

The bigger investment would be tying this into Ignite or others, but we can wait on that I think. Also note that we wouldn't need to include protos - those can just be referenced from TensorBoard itself.
